### PR TITLE
Add Doze Awareness Network Checking

### PR DESCRIPTION
**Pull request by heidijinxujia**
Issue #870
Power optimization is introduced into Android after Marshmallow. When the user enables the power optimization in Settings, the network will be disabled when the device is in doze mode or the app is standby. In this case, the networking checking should be updated accordingly.

Check if the device is in doze mode by checking boolean return of PowerManager.isDeviceIdleMode()
Check if the app is standby by checking the boolean return of UsageStatsManager.isAppInactive()
This PR add doze mode support into the network checking and return AuthenticationException with error code ADALError.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION when detected.

### CODE REVIEW

**Comment by heidijinxujia** [comment]
Unit tests got passed on VSTS, but failed on Travis. Investigating the cause now.

File: adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java [hunk 1]
```diff
@@ -82,6 +88,68 @@ protected void setUp() throws Exception {
...
+        } catch (final Exception exception) {
+            assertTrue(exception instanceof AuthenticationException);
+            assertTrue(((AuthenticationException) exception).getCode() == ADALError.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION);
+        }
+    }
+
+    /**
+     * Acquire token users refresh token, but the client app is inactive.
+     */
+    @Test
```
Comment (iambmelt) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java]: Is this the proper annotation to use? (I honestly don't know -- there seems to be a mix of `@SmallTest`, `@Test`, and unannotated test methods: not sure which ought to be used and when 🤔

**Comment by heidijinxujia** [comment]
@iambmelt Thank you, addressed.

File: adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java [hunk 2]
```diff
@@ -46,6 +50,10 @@
 import javax.crypto.spec.PBEKeySpec;
 import javax.crypto.spec.SecretKeySpec;
 
+import static android.R.attr.minSdkVersion;
```
Comment (iambmelt) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java]: The following imports appear unused:

```java
import org.junit.Test;
import static android.R.attr.minSdkVersion;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
```

File: adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java [hunk 3]
```diff
@@ -82,6 +90,60 @@ protected void setUp() throws Exception {
...
+        authenticationRequest.setUserId(TEST_IDTOKEN_USERID);
+        final AcquireTokenSilentHandler acquireTokenSilentHandler = getAcquireTokenHandler(mockContext,
+                authenticationRequest, mockCache);
+
+        try {
+            acquireTokenSilentHandler.acquireTokenWithRefreshToken("refreshToken");
+            fail("Expect exception");
+        } catch (final Exception exception) {
+            assertTrue(exception instanceof AuthenticationException);
+            assertTrue(((AuthenticationException) exception).getCode() == ADALError.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION);
```
Comment (weijjia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java]: @danieldobalian FYI. We're adding new error codes, you probably want to update the error doc you were working on.

File: adal/src/main/java/com/microsoft/aad/adal/UsageStatsManagerWrapper.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import android.annotation.TargetApi;
+import android.app.usage.UsageStatsManager;
+import android.content.Context;
+import android.os.Build;
+
+/**
+ * Wrapper class for UsageStatsManager.
+ */
+
+public class UsageStatsManagerWrapper {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/UsageStatsManagerWrapper.java]: Make this a method in DefaultConnectionService

File: adal/src/main/java/com/microsoft/aad/adal/PowerManagerWrapper.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import android.annotation.TargetApi;
+import android.content.Context;
+import android.os.Build;
+import android.os.PowerManager;
+
+/**
+ * Wrapper class for PowerManager.
+ */
+
+public class PowerManagerWrapper {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/PowerManagerWrapper.java]: Make this a method in DefaultConnectionService

File: adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java (hunk 2 above)
Comment (heidijinxujia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java]: Addressed but import static android.R.attr.minSdkVersion; is used in Line 92

Comment (heidijinxujia) [inline adal/src/androidTest/java/com/microsoft/aad/adal/AcquireTokenSilentHandlerTest.java]: @iambmelt I addressed the comment. Could you approve this PR if everything looks good to you?

File: adal/src/main/java/com/microsoft/aad/adal/DefaultConnectionService.java [hunk 6]
```diff
@@ -45,6 +49,31 @@ public boolean isConnectionAvailable() {
...
+     *
+     * @return true if the device is API23 and one or both of the following is true: the device is in doze or the company
+     * portal is in standby, false otherwise.
+     */
+    @TargetApi(Build.VERSION_CODES.M)
+    public boolean isNetworkDisabledFromOptimizations() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            final UsageStatsManagerWrapper usageStatsManagerWrapper = UsageStatsManagerWrapper.getInstance();
+            if (usageStatsManagerWrapper.isAppInactive(mConnectionContext)) {
+                Logger.w(TAG, "Company Portal is Inactive. Network is disabled.", "", ADALError.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION);
```
Comment (weijjia) [inline adal/src/main/java/com/microsoft/aad/adal/DefaultConnectionService.java]: why the log message is "company portal is inactive"?

**Comment by heidijinxujia** [comment]
@addev-ashish @weijjia I have addressed the comments. Thanks.

**Pull request merged**